### PR TITLE
Client State part2, Consensus State, Header and Misbehaviour

### PR DESCRIPTION
**Pull request by chist100**
Implementation of Client State p2, Header and Misbehaviour

### CODE REVIEW

File: modules/light-clients/14-avalanche/misbehaviour.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	// Ensure that Height1 is greater than or equal to Height2
+	if misbehaviour.Header1.SubnetHeader.Height.LT(misbehaviour.Header2.SubnetHeader.Height) {
+		return errorsmod.Wrapf(clienttypes.ErrInvalidMisbehaviour, "Header1 height is less than Header2 height (%s < %s)", misbehaviour.Header1.SubnetHeader.Height, misbehaviour.Header2.SubnetHeader.Height)
+	}
+
+	return nil
+}
+
+// validCommit checks if the given commit is a valid commit from the passed-in validatorset
+func validCommit(chainID string, blockID tmtypes.BlockID, commit *tmproto.Commit, valSet *tmproto.ValidatorSet) (err error) {
```
Comment (ramilexe) [inline modules/light-clients/14-avalanche/misbehaviour.go]: no needs for this function

File: modules/light-clients/14-avalanche/misbehaviour.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+var _ exported.ClientMessage = &Misbehaviour{}
+
+// FrozenHeight is same for all misbehaviour
+var FrozenHeight = clienttypes.NewHeight(0, 1)
+
+// NewMisbehaviour creates a new Misbehaviour instance.
+func NewMisbehaviour(clientID string, header1, header2 *Header) *Misbehaviour {
+	return &Misbehaviour{
+		ClientId: clientID,
```
Comment (ramilexe) [inline modules/light-clients/14-avalanche/misbehaviour.go]: It seems ClientID is no used in the code. Please remove from proto and here

**Review by ramilexe — CHANGES_REQUESTED** [review]
Please check comments

File: modules/light-clients/14-avalanche/client_state_test.go [hunk 3]
```diff
@@ -194,3 +210,876 @@ func (suite *AvalancheTestSuite) TestInitialize() {
 		}
 	}
 }
+
```
Comment (KyleMoser) [inline modules/light-clients/14-avalanche/client_state_test.go:213]: Some of the tests in the tendermint client for `TestVerifyMembership` aren't here, like "successful ConsensusState verification", "successful Connection verification", etc. Do we need those?

File: modules/light-clients/14-avalanche/client_state.go [hunk 4]
```diff
@@ -283,8 +342,335 @@ func IsMatchingClientState(subject, substitute ClientState) bool {
...
+		return err
+	}
+
+	chainID, _ := ids.ToID([]byte(cs.ChainId))
+	unsignedMsg, _ := warp.NewUnsignedMessage(
+		chainID,
+		ids.Empty,
+		consensusState.ValidatorSet,
+	)
+
```
Comment (KyleMoser) [inline modules/light-clients/14-avalanche/client_state.go:436]: Can we add some comments here, please? Like "Check that 2/3rds of validators signed validator set" (simple statement of what we are verifying). Same for the storage root `Verify()` call which, from our discussions, I understand to be the signed merkle root used for proof verification.

File: modules/light-clients/14-avalanche/update.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+// VerifyClientMessage checks if the clientMessage is of type Header or Misbehaviour and verifies the message
+func (cs *ClientState) VerifyClientMessage(
+	ctx sdk.Context, cdc codec.BinaryCodec, clientStore sdk.KVStore,
+	clientMsg exported.ClientMessage,
+) error {
+	switch msg := clientMsg.(type) {
+	case *Header:
+		err := cs.verifyHeader(ctx, clientStore, cdc, msg)
+		if err != nil {
```
Comment (KyleMoser) [inline modules/light-clients/14-avalanche/update.go]: Assume this was from debugging, but suggest changing this to `return cs.verifyHeader(ctx, clientStore, cdc, msg)`

File: modules/light-clients/14-avalanche/client_state.go [hunk 6]
```diff
@@ -181,33 +192,81 @@ func verifyDelayPeriodPassed(ctx sdk.Context, store sdk.KVStore, proofHeight exp
...
-func (cs *ClientState) CheckForMisbehaviour(ctx sdk.Context, cdc codec.BinaryCodec, clientStore sdk.KVStore, msg exported.ClientMessage) bool {
-	//TODO implement me
-	panic("implement me")
-}
 
-func (cs *ClientState) VerifyNonMembership(ctx sdk.Context, clientStore sdk.KVStore, cdc codec.BinaryCodec, height exported.Height, delayTimePeriod uint64, delayBlockPeriod uint64, proof []byte, path exported.Path) error {
-	//TODO implement me
-	panic("implement me")
-}
+func (cs *ClientState) UpdateStateOnMisbehaviour(ctx sdk.Context, cdc codec.BinaryCodec, clientStore sdk.KVStore, _ exported.ClientMessage) {
```
Comment (KyleMoser) [inline modules/light-clients/14-avalanche/client_state.go]: Suggest moving this to update.go to be consistent with the tendermint light client. Same with the UpdateState function.

File: modules/light-clients/14-avalanche/client_state.go [hunk 7]
```diff
@@ -283,8 +342,335 @@ func IsMatchingClientState(subject, substitute ClientState) bool {
 	return reflect.DeepEqual(subject, substitute)
 }
 
-func (cs *ClientState) VerifyUpgradeAndUpdateState(ctx sdk.Context, cdc codec.BinaryCodec, store sdk.KVStore, newClient exported.ClientState, newConsState exported.ConsensusState, proofUpgradeClient, proofUpgradeConsState []byte) error {
-	//TODO implement me
+func (cs *ClientState) VerifyUpgradeAndUpdateState(ctx sdk.Context, cdc codec.BinaryCodec, clientStore sdk.KVStore, upgradedClient exported.ClientState, upgradedConsState exported.ConsensusState, proofUpgradeClient, proofUpgradeConsState []byte) error {
```
Comment (KyleMoser) [inline modules/light-clients/14-avalanche/client_state.go:294]: Recommend refactoring this function to rely on `VerifyMembership` similar to how the tendermint light client does.

File: modules/light-clients/14-avalanche/client_state.go (hunk 6 above)
Comment (chist100) [reply modules/light-clients/14-avalanche/client_state.go]: updated

File: modules/light-clients/14-avalanche/client_state.go (hunk 7 above)
Comment (chist100) [reply modules/light-clients/14-avalanche/client_state.go:294]: updated

File: modules/light-clients/14-avalanche/update.go (hunk 5 above)
Comment (chist100) [reply modules/light-clients/14-avalanche/update.go]: updated

File: modules/light-clients/14-avalanche/client_state.go (hunk 6 above)
Comment (chist100) [reply modules/light-clients/14-avalanche/client_state.go]: update

File: modules/light-clients/14-avalanche/client_state.go (hunk 7 above)
Comment (chist100) [reply modules/light-clients/14-avalanche/client_state.go:294]: update

File: modules/light-clients/14-avalanche/client_state.go (hunk 4 above)
Comment (chist100) [reply modules/light-clients/14-avalanche/client_state.go:436]: updated

**Review by KyleMoser — APPROVED** [review]

**Pull request merged**